test/common/stats/lazyinit_stats_test.cc

```diff
@@ -0,0 +1,177 @@
+#include "envoy/stats/scope.h"
+#include "envoy/stats/stats_macros.h"
+
+#include "source/common/common/thread.h"
+#include "source/common/stats/lazy_init.h"
+#include "source/common/stats/thread_local_store.h"
+
+#include "test/test_common/utility.h"
+
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+namespace Envoy {
+namespace Stats {
+namespace {
+
+#define AWESOME_STATS(COUNTER, GAUGE, HISTOGRAM, TEXT_READOUT, STATNAME) COUNTER(foo)
+
+MAKE_STAT_NAMES_STRUCT(AwesomeStatNames, AWESOME_STATS);
+MAKE_STATS_STRUCT(AwesomeStats, AwesomeStatNames, AWESOME_STATS);
+
+class LazyInitStatsTest : public testing::Test {
+public:
+  SymbolTableImpl symbol_table_;
+  AllocatorImpl allocator_{symbol_table_};
+  ThreadLocalStoreImpl store_{allocator_};
+  AwesomeStatNames stats_names_{symbol_table_};
+};
+
+using MyStats = LazyCompatibleStats<AwesomeStats>;
+
+// Tests that "AwesomeStats.initialized" gauge equals the number of initiated MyStats instances.
+TEST_F(LazyInitStatsTest, StatsGoneWithScope) {
+  {
+    ScopeSharedPtr scope = store_.createScope("bluh");
+    // No such gauge when there is no lazy init stats instances.
+    EXPECT_EQ(TestUtility::findGauge(store_, "bluh.AwesomeStats.initialized"), nullptr);
+    MyStats x = MyStats::create(scope, stats_names_, true);
+    MyStats y = MyStats::create(scope, stats_names_, true);
+    EXPECT_EQ(TestUtility::findGauge(store_, "bluh.AwesomeStats.initialized")->value(), 0);
+    x->foo_.inc();
+    EXPECT_EQ(TestUtility::findGauge(store_, "bluh.AwesomeStats.initialized")->value(), 1);
+    EXPECT_EQ(x->foo_.value(), 1);
+    EXPECT_EQ(y->foo_.value(), 1);
+    EXPECT_EQ(TestUtility::findGauge(store_, "bluh.AwesomeStats.initialized")->value(), 2);
+  }
+  // Deleted as scope deleted.
+  EXPECT_EQ(TestUtility::findCounter(store_, "bluh.foo"), nullptr);
+  EXPECT_EQ(TestUtility::findGauge(store_, "bluh.AwesomeStats.initialized"), nullptr);
+  {
+    // Recreate scope "bluh".
+    ScopeSharedPtr scope = store_.createScope("bluh");
+    EXPECT_EQ(TestUtility::findGauge(store_, "bluh.AwesomeStats.initialized"), nullptr);
+    MyStats x = MyStats::create(scope, stats_names_, true);
+    EXPECT_EQ(TestUtility::findGauge(store_, "bluh.AwesomeStats.initialized")->value(), 0);
+    // Previous data is gone, as scope_v2 and scope_1's lifecycle do not overlap.
+    EXPECT_EQ(x->foo_.value(), 0);
+    // Initialized now.
+    EXPECT_EQ(TestUtility::findGauge(store_, "bluh.AwesomeStats.initialized")->value(), 1);
+  }
+}
+
+// Tests that multiple stats struct instances within the same scope has no issue to keep the stats,
+// with removals.
+TEST_F(LazyInitStatsTest, MultipleInstancesSameScopeDynamicallyDestructed) {
+  {
+    ScopeSharedPtr scope_1 = store_.createScope("bluh");
+    auto x = std::make_unique<MyStats>(MyStats::create(scope_1, stats_names_, true));
```

Comment:
this looks strange; how does this work? MyStats::create() returns a MyStats, not a MyStats*.

Reply:
it's a move assign I believe, optimized by RVO.